doc/read-the-docs-site/doc/marconi-as-a-library/tutorials/write-first-indexer.rst

```diff
@@ -1,6 +1,105 @@
+.. _my-first-indexer:
+
 Write your first indexer
 ========================
 
 In this tutorial, we'll show you how to write the simplest possible indexer, and how to expose the query interface of your indexer from a JSON-RPC server.
 
-Coming eventually!
+In order to understand the tutorial, you need to understand:
+
+* Cardano blockchain (specifically `cardano-node`)
+* Haskell (mainly the cardano-api, lenses, sqlite-simple packages)
+* how chain-indexing in general
+
+The main building blocks of an indexer are:
+
+1) the event, which is the information you want to extract from a block and store it
+2) the instanciation of the backend
+3) the instanciation of the worker
+4) the query interface
+
+For reference, the import list is:
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART import
+   :end-before: BLOCKEND import
+
+Now, let's start with the definition of the event.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART event
+   :end-before: BLOCKEND event
+
+Given this event, we can instanciate the backend.
+Let's start with the simplest one: the ``ListIndexer``.
+The ``ListIndexer`` backend simply stores each event inside a ``Data.List``, and it is simply instanciated with ``Core.mkListIndexer``.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART ListIndexer
+   :end-before: BLOCKEND ListIndexer
+
+Once we have instanciated the backend with event type (``BlockInfoEvent``), we can wrap the indexer inside a ``Core.WorkerIndexer`` which is used to spin up a concurrent thread when reading blocks from the local Cardano node using the chain-sync protocol.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART ListIndexer worker
+   :end-before: BLOCKEND ListIndexer worker
+
+We now have a complete indexer which can be run with ``Core.runIndexer``.
+Next, let's start by defining a runtime environment type which will contain all necessary inputs to run the indexer:
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART env
+   :end-before: BLOCKEND env
+
+With that environment type, let's create the function which runs the indexer.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART runBlockInfoIndexer
+   :end-before: BLOCKEND runBlockInfoIndexer
+
+Putting the final pieces together, we get:
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART runBlockInfoListIndexer
+   :end-before: BLOCKEND runBlockInfoListIndexer
+
+After running that, you should get an output that looks like::
```

Comment:
Since running this depends on having a local node running, I don't think executing this tutorial is trivial for a newcomer. There should be some more details here on how to run this.

Reply:
Added more details here. Can you revalidate? https://github.com/input-output-hk/marconi/pull/202/files#diff-65f51defa37a1e7a1df25d2a0b28a13b463c7f731405616a27987ee61b66d874R73